Lib/glyphsLib/builder/instances.py

```diff
@@ -326,15 +327,17 @@ def set_width_class(ufo, designspace, instance):
     _set_class_from_instance(ufo, designspace, instance, WIDTH_AXIS_DEF)
 
 
-def apply_instance_data(designspace):
+# DEPRECATED: needs better API
```

Comment:
@belluzj what do you propose to use as an alternative?

Reply:
I don't think there is a working replacement now so maybe I should not mark this as deprecated but only as "todo"? The idea behind the deprecated is:

I think the `apply_instance_data` should work as independently from glyphsLib as possible. I marked the current function as "deprecated" because it takes as a parameter something that is produced by glyphsLib only, whereas I think it should take only a designspace and find everything it needs in it.

Actually, in the code here, the parameter "instance_data" is already only stuff from the designspace, but weirdly wrapped in something that looks like a dictionary to conform to expectations of existing users of the API. I think there should be a new API that more clearly exposes the intent, which is:

* inputs:
    * a designspace file that specifies instances
    * interpolated UFOs for those instances, at the locations pointed by each `instance.path` of the designspace
* outputs:
    * each existing UFO is updated according to special Glyphs.app customParameters/filters/whatever, whose values/configuration were found in the `instance.lib` of the designspace (and only there, no need to read the glyphs file again).

Also from a broader perspective I guess this feature of applying post-processing to interpolated instance UFOs will be needed as well in workflows that are not based on Glyphs, so I guess this function could be part of something else than glyphsLib, and implement postprocessing in a similar way as ufo2ft filters (the code of Glyphs.app-specific filters could be provided by glyphsLib though). It could also be done by MutatorMath or the "ufoProcessor" from the designspaceDocument repository, I'm not sure which is suitable.

Reply:
Thanks for the explanation. Yeah, I don't like the interface of that function for the same reasons, but any changes will need to be coordinated with fontmake.
It'd be nice to make this feature generic and non-glyphsLib specific. Probably mutatorMath (or ufoProcessor if/when it's fully feature-compatible with the latter) is a better place to define this API than is ufo2ft. The latter doesn't know about designspace yet, it's only concerned with converting UFOs to OTFs.